File: tests/pages/login_page.py
"""
Login Page Object.
Handles /login user authentication, form inputs, validation errors, and password reset.
"""

from playwright.sync_api import Page, expect
from tests.pages.base_page import BasePage


class LoginPage(BasePage):
    PATH = "/login"

    def __init__(self, page: Page):
        super().__init__(page)
        self.email_input = page.locator("input[type='email'], input[name='email']").first
        self.password_input = page.locator("input[type='password'], input[name='password']").first
        self.submit_button = page.locator("button[type='submit']").first
        self.error_banner = page.locator("[role='alert'], .text-red-600, .bg-red-50").first
        self.forgot_password_button = page.locator("button, a").filter(has_text="Forgot").first
        self.reset_email_input = page.locator("input[placeholder*='work email'], input[placeholder*='name@company.com'], input[type='email']").first
        self.send_reset_link_button = page.locator("button").filter(has_text="Send Reset Instructions").or_(page.locator("button").filter(has_text="Send")).first

    def login(self, email: str, password: str):
        """Fills credentials and submits the login form."""
        self.navigate()
        expect(self.email_input).to_be_visible()
        self.email_input.fill(email)
        self.password_input.fill(password)
        self.submit_button.click()
        return self

    def get_error_text(self) -> str:
        """Returns the displayed error message text."""
        banner = self.page.locator("[role='alert'], div.text-red-600, div.bg-red-50, p.text-red-600").first
        if banner.count() > 0:
            return banner.text_content() or ""
        return ""

    def open_forgot_password(self):
        """Opens the forgot password sub-form."""
        if self.forgot_password_button.is_visible():
            self.forgot_password_button.click()
            self.page.wait_for_timeout(300)

    def request_password_reset(self, email: str):
        """Requests password reset for the specified email."""
        self.navigate()
        self.open_forgot_password()
        if self.reset_email_input.is_visible():
            self.reset_email_input.fill(email)
        if self.send_reset_link_button.is_visible():
            self.send_reset_link_button.click()
        self.page.wait_for_timeout(500)

File: e2e/pages/LoginPage.ts
import { Page, Locator, expect } from "@playwright/test";
import { BasePage } from "./BasePage";
import { DEFAULT_PASSWORD } from "../fixtures/test-data";

export class LoginPage extends BasePage {
  readonly emailInput: Locator;
  readonly passwordInput: Locator;
  readonly submitButton: Locator;
  readonly errorBanner: Locator;
  readonly togglePasswordButton: Locator;
  readonly forgotPasswordButton: Locator;

  constructor(page: Page, baseURL?: string) {
    super(page, baseURL);
    this.emailInput = page.locator('input[data-testid="login-email"], input[type="email"], input[name="email"]').first();
    this.passwordInput = page.locator('input[data-testid="login-password"], input[type="password"], input[name="password"]').first();
    this.submitButton = page.locator('button[data-testid="login-submit"], button[type="submit"]').first();
    this.errorBanner = page.locator('[data-testid="login-error"], [role="alert"]').first();
    this.togglePasswordButton = page.locator('button[aria-label*="password" i], button[title*="password" i]').first();
    this.forgotPasswordButton = page.locator('button:has-text("Forgot")').first();
  }

  async goto(): Promise<void> {
    await this.navigate("/login");
  }

  async login(email: string, password: string = DEFAULT_PASSWORD): Promise<void> {
    await this.emailInput.click();
    await this.emailInput.fill(email);
    await expect(this.emailInput).toHaveValue(email);

    await this.passwordInput.click();
    await this.passwordInput.fill(password);
    await expect(this.passwordInput).toHaveValue(password);

    await this.submitButton.click();
  }

  async assertErrorVisible(errorMatch?: string | RegExp): Promise<void> {
    await expect(this.errorBanner).toBeVisible({ timeout: 8000 });
    if (errorMatch) {
      await expect(this.errorBanner).toContainText(errorMatch);
    }
  }

  async assertOnLoginPage(): Promise<void> {
    await expect(this.page).toHaveURL(/.*\/login.*/);
  }
}
